pkg/pillar/cmd/zedmanager/handlevolumemgr.go

```diff
@@ -12,167 +12,157 @@ import (
-		log.Infof("VolumeConfig exists for %s to refcount %d",
-			key, m.RefCount)
-		publishVolumeConfig(ctx, m)
+		m.RefCount += vrc.RefCount
+		log.Infof("VolumeRefConfig exists for %s to refcount %d",
+			vrc.Key(), m.RefCount)
+		publishVolumeRefConfig(ctx, m)
 	} else {
-		log.Debugf("AddOrRefcountVolumeConfig: add for %s", key)
-		// XXX hard-coded for OriginTypeDownload for now
-		d := types.DownloadOriginConfig{
-			ImageID:     ss.ImageID,
-			DatastoreID: ss.DatastoreID,
-			Name:        ss.Name,
-			NameIsURL:   ss.NameIsURL,
-			ImageSha256: ss.ImageSha256,
-			IsContainer: ss.IsContainer,
-			AllowNonFreePort: types.AllowNonFreePort(*ctx.globalConfig,
-				types.AppImgObj),
-			MaxDownSize: ss.MaxDownSize,
-			FinalObjDir: ss.FinalObjDir, // XXX use?
-
-			CertificateChain: ss.CertificateChain,
-			ImageSignature:   ss.ImageSignature,
-			SignatureKey:     ss.SignatureKey,
-		}
-		n := types.OldVolumeConfig{
-			BlobSha256:     blobSha256,
-			AppInstID:      appInstID,
-			VolumeID:       volumeID,
-			PurgeCounter:   purgeCounter,
-			DisplayName:    ss.Name,
-			Origin:         types.OriginTypeDownload,
-			DownloadOrigin: &d,
-			MaxVolSize:     ss.MaxVolSize,
-			ReadOnly:       ss.ReadOnly,
-			Format:         ss.Format,
-			Target:         ss.Target,
-			RefCount:       1,
-		}
-		publishVolumeConfig(ctx, &n)
+		log.Debugf("MaybeAddVolumeRefConfig: add for %s", vrc.Key())
+		publishVolumeRefConfig(ctx, &vrc)
 	}
 	base.NewRelationObject(base.AddRelationType, base.AppInstanceConfigLogType, appInstID.String(),
-		base.OldVolumeConfigLogType, key).Infof("App instance to volume relation.")
-	log.Infof("AddOrRefcountVolumeConfig done for %s", key)
+		base.VolumeRefConfigLogType, vrc.Key()).Infof("App instance to volume relation.")
+	log.Infof("MaybeAddVolumeRefConfig done for %s", vrc.Key())
 }
 
-// MaybeRemoveVolumeConfig decreases the RefCount and deletes the VolumeConfig
+// MaybeRemoveVolumeRefConfig decreases the RefCount and deletes the VolumeRefConfig
 // when the RefCount reaches zero
-func MaybeRemoveVolumeConfig(ctx *zedmanagerContext, blobSha256 string,
-	appInstID uuid.UUID, volumeID uuid.UUID, purgeCounter uint32) {
+func MaybeRemoveVolumeRefConfig(ctx *zedmanagerContext, appInstID uuid.UUID,
+	vrs types.VolumeRefStatus) {
 
-	key := types.VolumeKeyFromParts(blobSha256, appInstID, volumeID, purgeCounter)
-	log.Infof("MaybeRemoveVolumeConfig for %s", key)
-	m := lookupVolumeConfig(ctx, key)
+	log.Infof("MaybeRemoveVolumeRefConfig for %s", vrs.Key())
+	m := lookupVolumeRefConfig(ctx, vrs.Key())
 	if m == nil {
-		log.Infof("MaybeRemoveVolumeConfig: config missing for %s", key)
+		log.Infof("MaybeRemoveVolumeRefConfig: config missing for %s", vrs.Key())
 		return
 	}
 	if m.RefCount == 0 {
-		log.Fatalf("MaybeRemoveVolumeConfig: Attempting to reduce "+
-			"0 RefCount for %s", key)
+		log.Fatalf("MaybeRemoveVolumeRefConfig: Attempting to reduce "+
+			"0 RefCount for %s", vrs.Key())
 	}
 	m.RefCount--
 	if m.RefCount == 0 {
```

Comment:
This doesn't match the vrc.RefCount add above since that can be more than one. But can it really? Each VolumeRefConfig would only count as one??

Reply:
The logic of publishing VolumeRefConfig to the volumemgr is missing. RefCount will be greater than one in case of shared volume by the multiple app instances.